### PR TITLE
txn: Fix check_txn_status doesn't write rollback when key not exist

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

###  What have you changed?

The new `check_txn_status` didn't write rollback record when the transaction doesn't exist. This may because prewrite request on primary arrives later than those on secondaries. Once `check_txn_status` returns, the client will cleanup secondaries; and since `check_txn_status` didn't write rollback record, the later prewrite request may success and the transaction may commit. And since the secondary locks may be removed, the data may lose consistency.

This PR fixes this bug.

###  What is the type of the changes?

- Bugfix (a change which fixes an issue)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: src/storage/mvcc/txn.rs [hunk 1]
```diff
@@ -679,10 +679,20 @@ impl<S: Snapshot> MvccTxn<S> {
                     .reader
                     .get_txn_commit_info(&primary_key, self.start_ts)?
                 {
-                    Some((ts, write_type)) if write_type != WriteType::Rollback => {
-                        Ok((0, ts, false))
+                    Some((ts, write_type)) => {
+                        if write_type == WriteType::Rollback {
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: Cleanup has some metrics here. Do we need them as well here?

File: src/storage/mvcc/txn.rs [hunk 2]
```diff
@@ -679,10 +679,20 @@ impl<S: Snapshot> MvccTxn<S> {
...
+                            Ok((0, 0, false))
+                        } else {
+                            Ok((0, ts, false))
+                        }
+                    }
+                    None => {
+                        // Returning 0, 0 will allow the client to clean secondary locks. Put a
+                        // rollback record here to avoid it being successfully prewritten.
+                        let write = Write::new(WriteType::Rollback, self.start_ts, None);
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: Need to collapse rollback

**Comment by coocood** [comment]
LGTM

**Comment by MyonKeminta** [comment]
@sticnarf PTAL again

File: src/storage/mvcc/txn.rs [hunk 3]
```diff
@@ -697,10 +697,24 @@ impl<S: Snapshot> MvccTxn<S> {
                         }
                     }
                     None => {
-                        // Returning 0, 0 will allow the client to clean secondary locks. Put a
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: Maybe still keep this sentence of comment?

Comment (MyonKeminta) [reply src/storage/mvcc/txn.rs]: I think the new comments already says what this sentence wants to say. By the way I found that cleanup and check_txn_status looks very similar. I  think I'll try to simplify them in another PR.

**Comment by sticnarf** [comment]
LGTM. But I have another tiny problem. Currently we notify the lock manager if lock_ttl == 0 and commit_ts == 0. While there are still two cases when the condition is satisfied but there is no lock unlocked. Is it necessary to return another bool to indicate if a lock is unlocked?

File: src/storage/mvcc/txn.rs [hunk 4]
```diff
@@ -689,10 +689,34 @@ impl<S: Snapshot> MvccTxn<S> {
...
+                        } else {
+                            Ok((0, ts, false))
+                        }
+                    }
+                    None => {
+                        let ts = self.start_ts;
+
+                        // collapse previous rollback if exist.
+                        if self.collapse_rollback {
+                            self.collapse_prev_rollback(key.clone())?;
```
Comment (sticnarf) [inline src/storage/mvcc/txn.rs]: key -> primary_key

**Comment by MyonKeminta** [comment]
@sticnarf For the transaction A that invokes check_txn_status on another transaction B, it doesn't matter whether the transaction B is rolled back or unlocked by itself, I think.

**Comment by sticnarf** [comment]
> @sticnarf For the transaction A that invokes check_txn_status on another transaction B, it doesn't matter whether the transaction B is rolled back or unlocked by itself, I think.

I mean when there isn't a lock on the primary key, we don't need to wake up the waiters on that key.

**Comment by MyonKeminta** [comment]
@sticnarf Oh, I see. Let's fix it later and make this PR only responsible for fixing the bug.

File: src/storage/mvcc/txn.rs [hunk 5]
```diff
@@ -689,10 +689,34 @@ impl<S: Snapshot> MvccTxn<S> {
...
-                        Ok((0, ts, false))
+                    Some((ts, write_type)) => {
+                        if write_type == WriteType::Rollback {
+                            Ok((0, 0, false))
+                        } else {
+                            Ok((0, ts, false))
+                        }
+                    }
+                    None => {
+                        let ts = self.start_ts;
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: Can we extract a `rollback_nonexistent_lock`? It's a repetition of `cleanup`.

Comment (MyonKeminta) [reply src/storage/mvcc/txn.rs]: Ok. I was thinking about simplifying it in another PR, but maybe I'd better do it right now.

Comment (youjiali1995) [reply src/storage/mvcc/txn.rs]: We can merge it first to let Jepsen pass.

**Pull request merged**